xcm/xcm-executor/src/assets.rs

```diff
@@ -198,14 +350,16 @@ impl Assets {
 					};
 					// remove the maxmimum possible up to id/amount from self, add the removed onto
 					// result
-					self.fungible.entry(id.clone())
-						.and_modify(|e| if *e >= amount {
+					let maybe_value = self.fungible.get(&id);
+					if let Some(&e) = maybe_value {
+						if e > amount {
+							self.fungible.insert(id.clone(), e - amount);
 							result.saturating_subsume_fungible(id, amount);
-							*e = *e - amount;
 						} else {
-							result.saturating_subsume_fungible(id, *e);
-							*e = 0
-						});
+							self.fungible.remove(&id);
+							result.saturating_subsume_fungible(id, e.clone());
+						}
+					}
```

Comment:
The old behavior would leave behind an asset with 0 amount.
This new logic removes an asset from the map if the value is zero.